pyfarm/master/api/jobgroups.py

```diff
@@ -0,0 +1,355 @@
+from pyfarm.models.jobtype import JobType
+from pyfarm.models.job import Job
+from pyfarm.models.jobgroup import JobGroup
+from pyfarm.models.core.cfg import  MAX_USERNAME_LENGTH, MAX_JOBTYPE_LENGTH
+from pyfarm.master.utility import jsonify, validate_with_model
+from pyfarm.master.application import db
+
+logger = getLogger("api.jobgroups")
+
+AUTOCREATE_USERS = read_env_bool("PYFARM_AUTOCREATE_USERS", True)
+AUTO_USERS_DEFAULT_DOMAIN = read_env("PYFARM_AUTO_USERS_DEFAULT_DOMAIN", None)
+
+
+def schema():
+    """
+    Returns the basic schema of :class:`.JobGroup`
+
+    .. http:get:: /api/v1/jobgroups/schema HTTP/1.1
+
+        **Request**
+
+        .. sourcecode:: http
+
+            GET /api/v1/jobgroups/schema HTTP/1.1
+            Accept: application/json
+
+        **Response**
+
+        .. sourcecode:: http
+
+            HTTP/1.1 200 OK
+            Content-Type: application/json
+
+            {
+                "main_jobtype": "VARCHAR(64)",
+                "title": "VARCHAR(255)",
+                "user": "VARCHAR(255)",
+                "id": "INTEGER"
+            }
+
+    :statuscode 200: no error
+    """
+    schema_dict = JobGroup.to_schema()
+    # In the database, we are storing the user by id, but over the wire, we are
+    # using the username to identify the user instead.
+    schema_dict["user"] = "VARCHAR(%s)" % MAX_USERNAME_LENGTH
+    del schema_dict["user_id"]
+    schema_dict["main_jobtype"] = "VARCHAR(%s)" % MAX_JOBTYPE_LENGTH
+    del schema_dict["main_jobtype_id"]
+    return jsonify(schema_dict), OK
+
+class JobGroupIndexAPI(MethodView):
+    @validate_with_model(JobGroup, ignore=["user", "main_jobtype"],
+                         disallow=["user_id", "main_jobtype_id"])
+    def post(self):
+        """
+        A ``POST`` to this endpoint will create a new job group.
+
+        .. http:post:: /api/v1/jobgroups/ HTTP/1.1
+
+            **Request**
+
+            .. sourcecode:: http
+
+                POST /api/v1/jobgroups/ HTTP/1.1
+                Accept: application/json
+
+                {
+                    "title": "Test Group",
+                    "user": "testuser",
+                    "main_jobtype": "Test JobType"
+                }
+
+
+            **Response**
+
+            .. sourcecode:: http
+
+                HTTP/1.1 201 CREATED
+                Content-Type: application/json
+
+                {
+                    "id": 2,
+                    "jobs": [],
+                    "user": "testuser",
+                    "main_jobtype": "Test JobType",
+                    "title": "Test Group"
+                }
+
+        :statuscode 201: a new job group was created
+        :statuscode 400: there was something wrong with the request (such as
+                         invalid columns being included)
+        """
+        username = g.json.pop("user")
+        user = User.query.filter_by(username=username).first()
+        if not user and AUTOCREATE_USERS:
```

Comment:
The pattern

    object = some query
    if not object:
        something else

is already being used almost everywhere in the codebase where we need something like this.